teps/0011-redirecting-step-output-streams.md

```diff
@@ -0,0 +1,212 @@
+
+### Goals
+
+1. Allow the stdout/stderr of a step to be consumed by another step.
+1. Enable a user to configure the path where the output streams are written.
+
+### Non-Goals
+
+1. Parse stdout/stderr of a step into a structured format (e.g., JSON) and extract information from certain fields.
+
+## Requirements
+
+* Add new fields for a step to specify paths to redirect stdout/stderr to.
+* Users should be able to observe the output streams through Pod logs even if `StdoutPath` or `StderrPath` is specified. In other words, output streams should be duplicated instead of simply redirected.
+* Clearly documents the restrictions when `StdoutPath` and `StderrPath` are set to Task result paths, and encourage Task authors to set the paths to workspace paths, especially if they want to exchange large amount of data between Tasks.
+* Provide examples to use step redirection for the use cases mentioned in [Motivation](#motivation).
+
+## Proposal
+
+### Option 1: Allowing Users to Specify Redirection Paths
+
+The following new fields will be added to the `Step` struct:
+
+```go
+type Step struct {
+    ...
+    // Path to duplicate stdout stream to on container's local filesystem.
+    // +optional
+    StdoutPath string `json:"stdoutPath,omitempty"`
+    // Path to duplicate stderr stream to on container's local filesystem.
+    // +optional
+    StderrPath string `json:"stderrPath,omitempty"`
+}
+```
+
+Once `StdoutPath` or `StderrPath` is specified, the corresponding output stream will be duplicated to both the given file and the standard output stream of the container, so users can still view the output through the Pod log API. If both `StdoutPath` and `StderrPath` are set to the same value, the two streams will be written to two file descriptors referring to the same file description, so outputs from both stream will be interleaved in the same file, but there will be no ordering guarantee on the data.  If multiple steps' `StdoutPath` fields are set to the same value, the file content will be overwritten by the last outputting step.
+
+Variable substitution will be applied to the new fields, so one could specify `$(results.<name>.path)` to the `StdoutPath` field to extract the stdout of a step into a Task result. No new variable substitution for accessing the values of `StdoutPath` and `StderrPath` fields will be provided so variable substitution can remain single-pass.
+
+#### Example Usage
+
+Redirecting stdout of `boskosctl` to `jq` and publish the resulting `project-id` as a Task result:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: boskos-acquire
+spec:
+  results:
+  - name: project-id
+  steps:
+  - name: boskosctl
+    image: gcr.io/k8s-staging-boskos/boskosctl
+    args:
+    - acquire
```

Comment:
I have a relatively stupid question (sorry :D), shouldn't this be possible today?

```yaml
image: gcr.io/k8s-staging-boskos/boskosctl
command:
  - /bin/boskosctl --server-url=http://boskos.test-pods.svc.cluster.local --owner-name=christie-test-boskos --type=gke-project --state=free --target-state=busy 1>/data/boskosctl-stdout
```

As far as I am aware, you can do redirection without a shell, you just need to execute your entire command in one string that is passed to the container runtime instead of using the args section.

Reply:
@pierretasci nop, this is not possible. *redirection* (`1>/data/boskosctl-stdout`) is a shell feature. When you use `command`, the process is not started in a shell (by kubernetes), so `1>/data/boskosctl-stdout` would be passed as argument to `/bin/boskosctl`. This is not even kubernetes specific, but all container runtime do the same : `docker` with an entrypoint that is `/bin/boskosctl` doesn't support redirection, because no shell is involved.

Reply:
Again, not the expert here...but https://docs.docker.com/engine/reference/builder/#cmd

> CMD command param1 param2 (shell form)
> If you want shell processing then either use the shell form or execute a shell directly

This is probably very CRI dependent though.

Reply:
@pierretasci this is the `Dockerfile` docs, *which have little to do with how the image is run*. 
In addition, `CMD` in `Dockerfile` is `args` in the `ContainerSpec`. `ENTRYPOINT` in the `Dockerfile` is `command` in the `ContainerSpec`. If you specify `command` in the `ContainerSpec` it completely overrides the image `ENTRYPOINT`, and pass it directly (no shell, nothing) — If you specify `args` on the other hands (without `command`), it passes those as arguments to the image `ENTRYPOINT`, which can be anything, no guarantee still that there is a shell.
From a `ContainerSpec`, the only way you can assume there is a shell available is if you know the image has one *and* you run it explicitely.